src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs

```diff
@@ -0,0 +1,3144 @@
+        for (int k = 0; k < 100; k++)
+        {
+            [|i|].Aggregate((m, n) => m + n);
+            j.Contains(100);
+            [|i|].Contains(100);
+        }
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(j As Integer())
+            Dim i = Enumerable.Range(1, 10)
+            For index = 1 To 10
+                [|i|].Aggregate(Function(m, n) m + n)
+                j.Contains(100)
+                [|i|].Contains(100)
+            Next
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestInvocationsInForEachLoop()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> j)
+    {
+        var i = Enumerable.Range(1, 10).ToArray();
+        foreach (var c in Enumerable.Range(1, 10))
+        {
+            i.Count();
+            [|j|].DefaultIfEmpty();
+        }
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(j As Integer())
+            Dim i = Enumerable.Range(1, 10)
+            For Each index in Enumerable.Range(1, 10)
+                [|i|].Aggregate(Function(m, n) m + n)
+                j.Contains(100)
+            Next
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestInvocationsAfterForEachLoop()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub()
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        foreach (var c in [|i|])
+        {
+        }
+        [|i|].Count();
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo()
+            Dim i = Enumerable.Range(1, 10)
+
+            For Each index in [|i|]
+            Next
+            [|i|].Count()
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestInvocationsAfterForEachLoopForArray()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub()
+    {
+        var i = Enumerable.Range(1, 10).ToArray();
+        foreach (var c in i.Select(p => p))
+        {
+        }
+        i.Count();
```

Comment:
I wonder why we are not warning here, i believe it is Linq (IEnumerable) method not Array

Reply:
(Why the comment shown as outdated)
`i` is an array here, is that correct?

Reply:
> (Why the comment shown as outdated)

hm strange all of them shows outdated, not sure why

> i is an array here, is that correct?

Yes but `Array` doesn't have `Count()` method right?

Reply:
Yes, my understanding it since `i` is an `Array` here, `Count()` won't call the GetEnuemrator() method to calculate the number of element, it will just fetch `Array.Count`, which is an O(1) operation. Is that correct?

Reply:
if you `ctrl+click` on the `Count` method of `i.Count();` above it will go to `IEnumerable.Count` extension method. As `Array` doesn't have `Count` method defined, and it implements `IEnumerable`, so when you call Array.Count it would call  `IEnumerable.Count` extension method

Reply:
Ah i see what you mean, even it calls `IEnumerable.Count` it would use ICollection.Cout(), never mind https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Count.cs#L18-L21